codec/enforceproto/extraneous_fields.go

```diff
@@ -0,0 +1,355 @@
+	5: {
+		descriptor.FieldDescriptorProto_TYPE_FIXED32:  true,
+		descriptor.FieldDescriptorProto_TYPE_SFIXED32: true,
+		descriptor.FieldDescriptorProto_TYPE_FLOAT:    true,
+	},
+}
+
+// canEncodeType returns true if the wireType is suitable for encoding the descriptor type.
+// See https://developers.google.com/protocol-buffers/docs/encoding#structure.
+func canEncodeType(wireType protowire.Type, descType descriptor.FieldDescriptorProto_Type) bool {
+	if iwt := int(wireType); iwt < 0 || iwt >= len(checks) {
+		return false
+	}
+	return checks[wireType][descType]
+}
+
+// errMismatchedWireType describes a mismatch between
+// expected and got wireTypes for a specific tag number.
+type errMismatchedWireType struct {
+	Type         string
+	GotWireType  protowire.Type
+	WantWireType protowire.Type
+	TagNum       protowire.Number
+}
+
+// String implements fmt.Stringer.
+func (mwt *errMismatchedWireType) String() string {
+	return fmt.Sprintf("Mismatched %q: {TagNum: %d, GotWireType: %q != WantWireType: %q}",
+		mwt.Type, mwt.TagNum, wireTypeToString(mwt.GotWireType), wireTypeToString(mwt.WantWireType))
+}
+
+// Error implements the error interface.
+func (mwt *errMismatchedWireType) Error() string {
+	return mwt.String()
+}
+
+var _ error = (*errMismatchedWireType)(nil)
+
+func wireTypeToString(wt protowire.Type) string {
+	switch wt {
+	case 0:
+		return "varint"
+	case 1:
+		return "fixed64"
+	case 2:
+		return "bytes"
+	case 3:
+		return "start_group"
+	case 4:
+		return "end_group"
+	case 5:
+		return "fixed32"
+	default:
+		return fmt.Sprintf("unknown type: %d", wt)
+	}
+}
+
+// errExtraneousField represents an error indicating that we encountered
+// a field that isn't available in the target proto.Message.
+type errExtraneousField struct {
+	Type     string
+	TagNum   protowire.Number
+	WireType protowire.Type
+}
+
+// String implements fmt.Stringer.
+func (twt *errExtraneousField) String() string {
+	return fmt.Sprintf("errExtraneousField %q: {TagNum: %d, WireType:%q}",
+		twt.Type, twt.TagNum, wireTypeToString(twt.WireType))
+}
+
+// Error implements the error interface.
+func (twt *errExtraneousField) Error() string {
+	return twt.String()
+}
+
+var _ error = (*errExtraneousField)(nil)
+
+var (
+	protoFileToDesc   = make(map[string]*descriptor.FileDescriptorProto)
+	protoFileToDescMu sync.RWMutex
+)
+
+// Invoking descriptor.ForMessage(proto.Message.(Descriptor).Descriptor()) is incredibly slow
+// for every single message, thus the need for a hand-rolled custom version that's performant and cacheable.
+func extractFileDesMessageDesc(desc descriptorIface) (*descriptor.FileDescriptorProto, *descriptor.DescriptorProto, error) {
+	gzippedPb, indices := desc.Descriptor()
+
+	protoFileToDescMu.RLock()
+	cached, ok := protoFileToDesc[string(gzippedPb)]
+	protoFileToDescMu.RUnlock()
+
+	if ok {
+		mdesc := cached.MessageType[indices[0]]
+		for _, index := range indices[1:] {
+			mdesc = mdesc.NestedType[index]
```

Comment:
I actually don't think we have a test for this case do we?

Reply:
I added one, thanks.